File: backend/src/AnimeShop/API/Security/EmployeeEntries.hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE TypeOperators #-}
module AnimeShop.API.Security.EmployeeEntries where

import AnimeShop.Prelude

type EmployeeEntriesAPI =
  "employee_entries" :>
  (
    Get '[JSON] [Entity EmployeeEntry] :<|>
    ReqBody '[JSON] EmployeeEntry :> Post '[JSON] (Entity EmployeeEntry) :<|>
    ReqBody '[JSON] EmployeeEntryId :> Delete '[JSON] () :<|>
    ReqBody '[JSON] (Entity EmployeeEntry) :> Patch '[JSON] ()
  )

employeeEntriesServer :: ServerT EmployeeEntriesAPI App
employeeEntriesServer = getEmployeeEntries :<|> postEmployeeEntry :<|> deleteEmployeeEntry :<|> patchEmployeeEntry

getEmployeeEntries :: App [Entity EmployeeEntry]
getEmployeeEntries = runDb $ selectList [] []

postEmployeeEntry :: EmployeeEntry -> App (Entity EmployeeEntry)
postEmployeeEntry = runDb . insertEntity

deleteEmployeeEntry :: EmployeeEntryId -> App ()
deleteEmployeeEntry = runDb . delete

patchEmployeeEntry :: Entity EmployeeEntry -> App ()
patchEmployeeEntry entity = runDb $ replace (entityKey entity) (entityVal entity)

File: frontend/src/reducers/security/employeeEntries.js
import { ADD_EMPLOYEE_ENTRY, GET_EMPLOYEE_ENTRIES, DELETE_EMPLOYEE_ENTRY } from '../../actions/types';

export function employeeEntries(state = [], action) {
  switch (action.type) {
    case ADD_EMPLOYEE_ENTRY:
      return [ ...state, action.employeeEntry];
    case GET_EMPLOYEE_ENTRIES:
      return action.employeeEntries;
    case DELETE_EMPLOYEE_ENTRY:
      return state.filter(employeeEntry => employeeEntry.id !== action.id);
    default:
      return state;
  }
}
